src/components/PeopleAndRoles/PeopleAndRoles.vue

```diff
@@ -189,40 +206,69 @@ export default class PeopleAndRoles extends Vue {
   }
 
   /**
-   * Sets properties to add an org/person.
+   * Determines whether we have the specified role by count and mode.
+   */
+  private hasRole (roleName: Roles, count: number, mode: Modes): boolean {
+    // 1. filter out removed people
+    // 2. filter in people with specified role
+    const orgPersonWithSpecifiedRole = this.getPeopleAndRoles
+      .filter(people => people.action !== ActionTypes.REMOVED)
+      .filter(people => people.roles.some(party => party.roleType === roleName))
+
+    if (mode === Modes.EXACT) {
+      return (orgPersonWithSpecifiedRole.length === count)
+    }
+    if (mode === Modes.AT_LEAST) {
+      return (orgPersonWithSpecifiedRole.length >= count)
+    }
+  }
+
+  /**
+   * Sets state properties to add an org/person.
    * @param roles The roles of this item.
    * @param type The incorporator (party) type of this item.
    */
-  private addOrgPerson (roles: RoleIF[], type: IncorporatorTypes): void {
-    this.currentOrgPerson = { ...this.emptyOrgPerson }
+  private initAdd (roles: RoleIF[], type: IncorporatorTypes): void {
+    // make a copy so we don't change the original object
+    this.currentOrgPerson = cloneDeep(this.emptyOrgPerson)
     this.currentOrgPerson.roles = roles
     this.currentOrgPerson.officer.partyType = type
     this.activeIndex = NaN
     this.nextId = (this.getPeopleAndRoles.length === 0)
-      ? 0 : this.getPeopleAndRoles[this.getPeopleAndRoles.length - 1].officer.id + 1
+      ? 0 : (this.getPeopleAndRoles[this.getPeopleAndRoles.length - 1].officer.id + 1)
     this.showOrgPersonForm = true
   }
 
   /**
-   * Sets properties to edit an org/person.
+   * Sets state properties to edit an org/person.
    * @param index The index of the org/person to edit.
    */
-  private editOrgPerson (index: number): void {
-    this.currentOrgPerson = { ...this.getPeopleAndRoles[index] }
+  private initEdit (index: number): void {
+    // make a copy so we don't change the original object
+    this.currentOrgPerson = cloneDeep(this.getPeopleAndRoles[index])
     this.activeIndex = index
     this.showOrgPersonForm = true
   }
 
+  /**
+   * Resets state properties after a change is completed (or to cancel).
+   */
+  private performReset (): void {
+    this.currentOrgPerson = null
+    this.activeIndex = NaN
+    this.showOrgPersonForm = false
+  }
+
   /**
    * Undoes changes to the specified org/person.
    * @param index The index of the org/person to undo.
    */
-  private undoOrgPerson (index: number): void {
-    // get org/person to undo
-    const person = this.getPeopleAndRoles[index]
-
+  private performUndo (index: number): void {
     // make a copy so Vue reacts when we set the updated list
-    const tempList = [...this.getPeopleAndRoles]
+    const tempList = cloneDeep(this.getPeopleAndRoles)
```

Comment:
OK this is critical...
Javascript spread (`...`) only does a shallow copy.
`Object.assign` only does a shallow copy.

Here, the top-level properties (ie, persons) were new objects, but the person's objects were references to the original objects. This means the code was actually changing the roles (or email) on the original object, corrupting future undos.

Reply:
very cool. Is `cloneDeep` part of a new package or did we have this already?

Reply:
https://lodash.com/docs/4.17.15#cloneDeep

But JS keeps getting richer and richer every year, so maybe one day we won't need Lodash.

I also considered writing my own (specific to OrgPersonIF properties) but it would need to be kept up to date when we extend the object. Another possibility is to create a copy constructor for this but I think that's overkill atm, eg,:
`const clonedList = new OrgPerson(originalList)`